### ISSUE TITLE
Can Enzyme differentiate data structure such as vector of vectors?

### ISSUE THREAD

**Issue by ipcamit** [EnzymeAD/Enzyme#585]
I of the programs I inherited in my lab has a data structure `std::vector<std::vector<double> >` , Although my program on compilation shows no error, during runtime it gives segmentation fault at precisely the line where a value from this structure is operated upon.

Hence I tried a minimal working example given below, but I could not get it to give correct results. Is it possible for Enzyme to handle such structures? Are there any documented known pitfalls I should be aware of? 

```c++
#include<iostream>
#include<vector>

void sum_all(double* mult, std::vector<std::vector<double> >& vector_vector, double * sum){
    *sum = 0.;
    for (auto vector: vector_vector){
        for (auto element: vector){
            *sum += element;
        }
    }
    *sum *= *mult;
}

void __enzyme_autodiff(void (*) (double *, std::vector<std::vector<double> >& , double * ), 
                                    /*duplicated argument*/double*, double*, 
                                    std::vector<std::vector<double> >& ,std::vector<std::vector<double> >& 
                                    , double *, double *);

int main(){
    std::vector<std::vector<double> > vector_vector(5), d_vector_vector(5);
    double sum = 0.; double d_sum = 0.;
    double mult = 2.; double d_mult = 0.;

    // initialize vector of vector
    for (int i = 0; i<5; i++){
        std::vector<double> vector(5,static_cast<double>(i));
        vector_vector[i] = std::move(vector);
        std::vector<double> vector2(5,0.0);
        d_vector_vector[i] = std::move(vector2);
    }

    __enzyme_autodiff(sum_all, &mult, &d_mult, vector_vector, d_vector_vector,  &sum, &d_sum);
    std::cout << d_mult <<"\n";
    return 0;
}
```

In above program I am getting d_mult = 0.0, it should be 50.0?

During compiling I do get warning about memmove not supported, and memcopy possibly giving error, but in my original gefaulting program no such warning is shown.

**Comment by tgymnich** [comment]
You forgot to initialise your output shadow (`d_sum`) with `1.0`.

**Comment by ipcamit** [comment]
Thank you. I think I got it now. One minor question that was not clear from the documentation, integers are by default `/*inactive argument*/` right? are `std::vector<int>` same or do they always need shadow?

Also is there any to get partial diff? As in for a function `f(double x, double y, double z, double result)` tell enzyme to only worry about operations directly involving x, and not worry about y, z?

**Comment by wsmoses** [comment]
Integers are by default inactive, std::vector<int> is a maybe, and I would probably put an enzyme_const in front of it, just in case.

Regarding partial diff, you can put enzyme_const in front of y, z, and you should get the desired behavior.

**Comment by ipcamit** [comment]
In the above example, how will `enzyme_const` work if I dont want to compute derivatives with respect to `<vector<vector>>`? 
Do comments like `/*inactive argument*/` give compiler any hints or are they for user only (I assumed that these comments were perhaps functioning as compiler paragma)? 

When I did try above example as per the example on documentation page, I got 
>  error: variable has incomplete type 'void'

```C++
#include<iostream>
#include<vector>

int enzyme_dup;
int enzyme_out;
int enzyme_const;

void sum_all(double* mult, std::vector<std::vector<double> >& vector_vector, double * sum){
    *sum = 0.;
    for (auto vector: vector_vector){
        for (auto element: vector){
            *sum += element;
        }
    }
    *sum *= *mult;
}

void __enzyme_autodiff(void (*) (double *, std::vector<std::vector<double> >& , double * ), 
                                    enzyme_dup, double*, double*, 
                                    enzyme_const, std::vector<std::vector<double> >&,
                                    enzyme_out, double *, double *);

int main(){
    std::vector<std::vector<double> > vector_vector(5), d_vector_vector(5);
    double sum = 0.; double d_sum = 1.;
    double mult = 2.; double d_mult = 0.;

    // initialize vector of vector
    for (int i = 0; i<5; i++){
        std::vector<double> vector(5,static_cast<double>(i));
        vector_vector[i] = std::move(vector);
        std::vector<double> vector2(5,0.0);
        d_vector_vector[i] = std::move(vector2);
    }

    __enzyme_autodiff(sum_all, enzyme_dup, &mult, &d_mult, enzyme_const,vector_vector, enzyme_out ,&sum, &d_sum);
    // __enzyme_autodiff(sum_all,  &mult, &d_mult, vector_vector, &sum, &d_sum);
    std::cout << d_mult <<"\n";
    return 0;
}
``` 
---
Sorry about any incessant or naive questioning! Our initial tests on simplified molecular dynamics system gave really encouraging results (benchmarks at the bottom), where we used AD systems to compute forces (F = -dEnergy/dr). Most AD systems were too slow for any serious scientific simulation therefore always need hand coded analytical functions. But Enzyme results were consistently with in 2x-3x time window of analytical functions at worst which means, we can provide AD calculated gradients for project [OpenKIM](https://openkim.org/). And I am tasked to port first few potentials and descriptors using Enzyme.


![E_F_C++](https://user-images.githubusercontent.com/6790415/159069689-3f297bff-8d45-4aea-93a3-8b5293968f53.png)

<small>In above graph Enzyme is actually performing better then analytical function because of some overhead, and the fact that analytical functions(labelled KIM) were called in python using pybind11. But this shows how close enzyme is to analytical derivatives, that small organizational overhead is enough to give it edge.</small>

**Comment by ZuseZ4** [comment]
These comments have now effect and should just show the defaults working there.
The following example shows how to overwrite those default expectations of Enzyme:
https://enzyme.mit.edu/getting_started/CallingConvention/#types

**Comment by wsmoses** [comment]
No worries, we love questions.

The error you saw was actually a C++ error, which can be fixed by doing two things. Having the int type used in your declaration of `__enzyme_autodiff` and also since you pass a pointer for the return, it should be marked `enzyme_dup`
```cpp
#include<iostream>
#include<vector>

int enzyme_dup;
int enzyme_out;
int enzyme_const;

void sum_all(double* mult, std::vector<std::vector<double> >& vector_vector, double * sum){
    *sum = 0.;
    for (auto vector: vector_vector){
        for (auto element: vector){
            *sum += element;
        }
    }
    *sum *= *mult;
}

void __enzyme_autodiff(void (*) (double *, std::vector<std::vector<double> >& , double * ), 
                                    int, double*, double*, 
                                    int, std::vector<std::vector<double> >&,
                                    int, double *, double *);

int main(){
    std::vector<std::vector<double> > vector_vector(5), d_vector_vector(5);
    double sum = 0.; double d_sum = 1.;
    double mult = 2.; double d_mult = 0.;

    // initialize vector of vector
    for (int i = 0; i<5; i++){
        std::vector<double> vector(5,static_cast<double>(i));
        vector_vector[i] = std::move(vector);
        std::vector<double> vector2(5,0.0);
        d_vector_vector[i] = std::move(vector2);
    }

    __enzyme_autodiff(sum_all, enzyme_dup, &mult, &d_mult, enzyme_const,vector_vector, enzyme_dup ,&sum, &d_sum);
    // __enzyme_autodiff(sum_all,  &mult, &d_mult, vector_vector, &sum, &d_sum);
    std::cout << d_mult <<"\n";
    return 0;
}
```

It's actually slightly better to just return sum, as below:

```cpp
#include<iostream>
#include<vector>

int enzyme_dup;
int enzyme_out;
int enzyme_const;

double sum_all(double* mult, std::vector<std::vector<double> >& vector_vector){
    double sum = 0.;
    for (auto vector: vector_vector){
        for (auto element: vector) {
            sum += element;
        }
    }
    return sum * *mult;
}

void __enzyme_autodiff(double (*) (double *, std::vector<std::vector<double> >&), 
                                    int, double*, double*, 
                                    int, std::vector<std::vector<double> >&
                                    );

int main(){
    std::vector<std::vector<double> > vector_vector(5), d_vector_vector(5);
    double sum = 0.; double d_sum = 1.;
    double mult = 2.; double d_mult = 0.;

    // initialize vector of vector
    for (int i = 0; i<5; i++){
        std::vector<double> vector(5,static_cast<double>(i));
        vector_vector[i] = std::move(vector);
        std::vector<double> vector2(5,0.0);
        d_vector_vector[i] = std::move(vector2);
    }

    __enzyme_autodiff(sum_all, enzyme_dup, &mult, &d_mult, enzyme_const,vector_vector);
    // __enzyme_autodiff(sum_all,  &mult, &d_mult, vector_vector, &sum, &d_sum);
    std::cout << d_mult <<"\n";
    return 0;
}
```

**Comment by ipcamit** [comment]
Thank you everyone, It really clarifies lot of confusion regarding different argument types.

**Comment by wsmoses** [comment]
Also since I see forward mode in your slide deck, thanks to @tgymnich Enzyme also supports (but does not yet have documentation for) forward mode AD as well.

If you have any good benchmarks for those, please send them our way as we'll presently looking for a good story to publish Forward Mode!

**Comment by ipcamit** [comment]
I havent yet benchmarked any FWD mode code. We were looking in forward mode because boost::autodiff works on fwd mode taylor expansion, and as boost is omnipresent, it would have saved us distribution headaches. But for above application of Coordinates(vector) to Energy(scalar), it was way too cumbersome to get it to work. However FWD mode enzyme will be perfect for our https://colabfit.org/ project where one of the application is coordinates to chemical descriptors (~10^4 dim vector). I would love to give it a try once FWD mode is released.

In fact if my current port works as intended I think we would definitely like to reach out to Enzyme in official capacity (PI to PI, I am but a lowly post-doc!). Fingers crossed! 

Another collaborator of ours is using currently JAX-MD, but ease of use and performance of Enzyme makes possible to couple it with LAMMPS(https://www.lammps.org/, which is already CLANG compatible). They are exploring that side (albeit slowly) but expect LAMMPS-Enzyme-MD to totally blow JAX-MD out of water.

**Comment by vchuravy** [comment]
Since I see that you tried out Zygote in Julia, I would be amiss to not mention that there is Enzyme.jl which uses Enzyme to differentiate Julia code. Features like BLAS Support and FFI Support are still in the works but it might be an interesting point of comparison as well.

**Comment by ipcamit** [comment]
Yes, nice point. Once I get little time I will try to benchmark Julia + Enzyme to see how well does that perform against Julia + Enzyme. 
The main reason julia is in that graph is that it was figure 2 in comparison of how good or bad AD systems can be for force evaluation. Figure 1 compared "mainstream" solutions like TorchScript and TensorFlow. It was to demonstrate that how conventional ML frameworks are not the best bet for Scientific applications. The model here is Stillinger-Weber potential for 64 Si atoms, which can not be optimized easily by TorchScript, and was so non-trivial to implement in JAX that I did not bother finishing it. Julia implementation provided a good benchmark to show how a good Language+AD model can work. See below the comparison

![E_F](https://user-images.githubusercontent.com/6790415/159096266-3eae46e0-df65-4860-8b6e-ed28d75e473d.png)

But given how good enzyme results came out to be, our focus shifted there after I showed these benchmarks. After I have finished porting some of these potentials and workflow to Enzyme AD, and benchmarked, next step for me would be to see If we can use it as the Torch backend (It was shown to work in the enzyme paper, but I havent seen any application in wild).
